ImplantFramework/lib/source/transports/linux_tcp.hpp

```diff
@@ -14,26 +17,121 @@
  */
 namespace hivemind_lib {
 
+#define DEFAULT_BUFLEN 1024
+
 /**
  * @brief LINUX implementation of the TCP transport.
  */
 class LinuxTcpTransport : public hivemind_lib::Transport {
  public:
   std::string SendAndReceive(std::string data) override {
-    return "";
+    std::string ret = "";
+    char buf[DEFAULT_BUFLEN];
+    int len = DEFAULT_BUFLEN;
+    int i_result;
+   
+    // Create socket
+    int socket = this -> ConnectTo();
+    if (socket < 0) {
+      DEBUG("ERROR WITH SOCKET", LEVEL_ERROR);
+      return "";
+    }
+
+    // Send data
+    i_result = write(socket, data.c_str(), strlen(data.c_str()));
+    if (i_result < 0) {
+      DEBUG("ERROR WRITING TO SOCKET", LEVEL_ERROR);
+      goto cleanup;
+    } else {
+      DEBUG("Bytes Sent:" << i_result, LEVEL_DEBUG);
+    }
+
+    // Shutdown sending
+    shutdown(socket, SHUT_WR);
+
+    // Receive data
+    do {
+      memset(buf, 0, sizeof(buf));
+      i_result = read(socket, buf, len);
+      if (i_result > 0) {
+        DEBUG("Bytes received: " << i_result, LEVEL_INFO);
+        ret += buf;
+      } else if (i_result == 0) {
+        DEBUG("Connection closed", LEVEL_INFO);
+      } else {
+        DEBUG("ERROR READING FROM SOCKET", LEVEL_ERROR);
+        goto cleanup;
+      }
+    } while (i_result > 0);
+
+    cleanup:
+      close(socket);
+      return ret;
   }
 
   void Send(std::string data) override {
+    int i_result;
+
+    // Create socket
+    int socket = this -> ConnectTo();
+    if (socket < 0) {
+      DEBUG("ERROR WITH SOCKET", LEVEL_ERROR);
+      goto cleanup;
```

Comment:
Not sure if calling close on a nonexisting socket handle will make the system very happy. We'll have to look out for this.

Reply:
It'll return -1 and set EBADF in errno. So we should just return here.